### DISCUSSION TITLE
Lock scenery button to prevent accidental demolishion while building rides

### DISCUSSION THREAD

**Discussion by FlyingThunder** [OpenRCT2/OpenRCT2#15804] [Feature requests]
Windows 10
0.3.2
cea5fab

Hello, i dont know if feature requests are in the right place here, but i would love to see a small button in the scenery menu that "locks" scenery objects, so you cant delete any while they are locked. So often while building a track, or trying to right-click a track piece, you accidently hit a scenery piece, which can be very frustrating if it happens in large scenery builds where you cant easily access and replace the accident.

- [x ] Reproducible in RCT2 (vanilla)?
- [x] Specific to multiplayer?

**Steps to reproduce:**
1. build rollercoaster close to scenery object and accidentally remove it when colliding with it
2. right click a ride, especially one with small hitbox like hairpin or wild mouse, and delete a path/scenery somewhere in the background

**Comment by IntelOrca** [comment]
This could probably be implemented with a plugin.

**Comment by duncanspumpkin** [comment]
I think this would be pretty hard to pull off. There are quite a few locations you would need to modify to add the extra checks.

**Comment by IntelOrca** [comment]
@duncanspumpkin all you need is a game action hook that returns an error for destroy scenery actions

**Comment by ocalhoun6** [comment]
I'm sure it would be _much_ more difficult to implement, but it seems like an Undo button would solve OP's problem much more elegantly ... and be useful in a lot of other situations as well. (ie, 'oops, I didn't mean to raise the ground there, I thought I was using the paintbrush mode' or 'oops, I didn't mean to demolish that ride'.)

If we're willing to deal with the limitation that undo history is not retained when you save and reload a game, then it could be implemented without waiting for a new file format.

And as difficult as it might be to actually accomplish, an undo button would be a _huge_ leap forward -- a major quality of life improvement over vanilla.

**Comment by duncanspumpkin** [comment]
@IntelOrca destroying scenery doesn't use game actions though map_place_clear_func is the function that does it. There are a few other functions which can do it as well scattered in other locations.

**Comment by IntelOrca** [comment]
OP only mentioned accidentally right clicking scenery. I assume you still want scenery cleared by building track to still happen.

**Comment by duncanspumpkin** [comment]
Oh I thought he was wanting both from the part where it talks about `So often while building a track,`. Yeah you could hook the game action.

**Comment by FlyingThunder** [comment]
i did mention both right clicking stuff and building rides into scenery as an example, but right clicking stuff is definetly happening a lot more than the later, and i agree that an undo button would be extremely nice in general

**Comment by duncanspumpkin** [comment]
Undo button is something we want to work eventually. Needs a good bit of work though.

**Comment by MASHzj** [comment]
Your second problem (accidentally removing scenery while right clicking on a ride) can be resolved by setting scenery and footpaths invisible instead of see-through with the eye-icon in Transparency options (accessible via View Options menu).
